orderer/common/multichannel/registrar.go

```diff
@@ -577,8 +577,71 @@ func (r *Registrar) joinAsFollower(ledgerRes *ledgerResources, clusterConsenter
 // RemoveChannel instructs the orderer to remove a channel.
 // Depending on the removeStorage parameter, the storage resources are either removed or archived.
 func (r *Registrar) RemoveChannel(channelID string, removeStorage bool) error {
+	r.lock.RLock()
+	defer r.lock.RUnlock()
+
 	if r.SystemChannelID() != "" && channelID != r.SystemChannelID() {
 		return types.ErrSystemChannelExists
 	}
-	return errors.New("Not implemented yet")
+
+	cs, ok := r.chains[channelID]
+	if ok {
+		return r.removeMember(channelID, cs)
+	}
+
+	follower, ok := r.followers[channelID]
+	if ok {
+		return r.removeFollower(channelID, follower)
+	}
+	return types.ErrChannelNotExist
+}
+
+func (r *Registrar) removeMember(channelID string, cs *ChainSupport) error {
+	err := validateRaftChain(channelID, cs)
+	if err != nil {
+		return err
+	}
+
+	cs.Halt()
+
+	err = r.ledgerFactory.Remove(channelID)
+	if err != nil {
+		return errors.Errorf("error removing ledger for channel %s", channelID)
+	}
+
+	delete(r.chains, channelID)
+
+	logger.Infof("Removed channel: %s", channelID)
+
+	return nil
+}
+
+func (r *Registrar) removeFollower(channelID string, follower *follower.Chain) error {
+	err := validateRaftChain(channelID, follower)
+	if err != nil {
+		return err
+	}
+
+	follower.Halt()
+
+	err = r.ledgerFactory.Remove(channelID)
+	if err != nil {
+		return errors.Errorf("error removing ledger for channel %s", channelID)
+	}
+
+	delete(r.followers, channelID)
+
+	logger.Infof("Removed channel: %s", channelID)
+
+	return nil
+}
+
+// validateRaftChain returns an error if the chain is not a raft chain.
+func validateRaftChain(channelID string, chain consensus.StatusReporter) error {
+	clusterRelation, _ := chain.StatusReport()
+	if clusterRelation == types.ClusterRelationNone {
+		return errors.Errorf("cannot remove non-raft application channel %s", channelID)
+	}
+
+	return nil
```

Comment:
I am thinking, there is nothing in the "Remove" that is specific to raft/follower... We call `Halt`, we remove the ledger, and that is it (unlike the "Join"...). So perhaps we can make the code simpler by removing this check?

Reply:
Done